Flash_Circular_Buffer basing on Flash_map module

nvlsianpu opened the pull request:
This patch objective is addition of Flash Circular Buffer module as zephyr native code.
Code of the Flash Circular Buffer is taken from MyNewt.
API of the flash_map module is almost the same as for MyNewt. Implementation of this part is mostly reworked for zephyr, flash_area_get_sectors API and its implementation is taken form mcuboot.

Explanation for choice addition of flash_map module:
Of course the providing of substitute of flash_area primitives is intermedium step for adding other modules. I will adapt mynewt FCB (Flash Circular Buffer) to zephyr, I will take mynewt Config module (or something very similar ) as well. What I also potentially want to do providing something like mynewt’s Manufacturer meta region is – this would be useful for inform both bootloader and application about how memory map looks like. I think that zephyr flash_map will simplify support for multi-memories and multi-images mcuboot support as well or event we can support dynamic image slots allocation basing on that (so will be possible to change flash areas boundaries during FWU – we see the case for that).

Question:
Should FCB module been moved to File System directory (so Kconfig as well)?

Should I provide better way for providing user's custom flash map description instead of CONFIG_FLASH_MAP_CUSTOM option (look at https://github.com/nvlsianpu/zephyr/blob/add_fcb/subsys/storage/flash_map/CMakeLists.txt#L2 ).

nvlsianpu committed:
subsys: storage: Add flash_map abstraction module

Introduce flas_map module is abstraction over flash memory and its
driver for using flash memories along with description of
available flash areas.
Module provides simple API for write/read/erase and so one.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
lib: crc: Add crc8-ccitt implementation

This patch adds crc8-ccitt calculation routine.

CRC8 CCITT is required for Flash Circular Buffer
module (originally mynewt module).

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
subsys: fs: Add Flash Circular Buffer

Zephyr already supports NFFS as a storage layer, but it might
be a little bit too heavyweight for certain applications in
memory-restricted ICs.

This module is response for need of Lightweight flash storage
capability. FCB is ported form MyNewt as native zephyr module.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: fs: Add tests Flash Circular Buffer module

Tests added were adopted from MyNewt.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

SebastianBoe commented on subsys/flash_map/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+zephyr_sources_ifdef(CONFIG_FLASH_MAP flash_map.c)
+
+if(CONFIG_FLASH_MAP)
+zephyr_sources_ifndef(CONFIG_FLASH_MAP_CUSTOM flash_map_default.c)
```
Indentation.

SebastianBoe replied:
You have alread ifdef'd this entire directory, so you dont' need to ifdef flash_map again here. The same applies to FCB I assume.

SebastianBoe commented on subsys/fcb/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+zephyr_sources_ifdef(CONFIG_FCB fcb_append.c  fcb.c  fcb_elem_info.c  fcb_getnext.c fcb_rotate.c  fcb_walk.c)
```
it is easier to read (and to diff) if each file is on it's own line.

nashif commented on include/crc8.h:
```diff
@@ -0,0 +1,28 @@
+#endif
+
+u8_t crc8_init(void);
+u8_t crc8_ccitt(u8_t val, void *buf, int cnt);
```
missing documentation of this API

nashif commented on drivers/crc/crc8_sw.c:
```diff
@@ -0,0 +1,31 @@
+/*
```
this file needs to go into lib/crc (crc has moved from drivers/)

nashif replied:
also please add tests for this new API

nashif commented on subsys/fcb/fcb.c:
```diff
@@ -0,0 +1,252 @@
+/*
```
Dont think this makes a subsystem on its own, this belongs rather into something else, maybe subsys/storage.

nashif replied:
or under filesystem... (fs)

carlescufi replied:
I suggest we place it under filesystem

nashif commented on samples/basic/blinky/prj.conf:
```diff
@@ -1,2 +1,3 @@
 CONFIG_GPIO=y
 CONFIG_SERIAL=n
+CONFIG_FLASH_MAP=y
```
why is this needed here?

nashif commented on subsys/flash_map/Kconfig:
```diff
@@ -0,0 +1,27 @@
+	default n
+	depends on FLASH_MAP
+	help
+	  User must provide custom flash map description in place of default on
```
rephrase this help message, something is missing.

nashif commented on subsys/flash_map/flash_map.c:
```diff
@@ -0,0 +1,270 @@
+/*
```
should not be in its own subsystem, this needs to be either part of fs or a new subsystem with generic name, like storage for example.

nashif commented on subsys/flash_map/flash_map.c:
```diff
@@ -0,0 +1,270 @@
+}
+
+int flash_area_write(const struct flash_area *fa, off_t off, const void *src,
+			 size_t len)
```
align

nashif commented on subsys/flash_map/flash_map.c:
```diff
@@ -0,0 +1,270 @@
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+int flash_area_read(const struct flash_area *fa, off_t off, void *dst,
+			size_t len)
```
align

nashif commented on tests/subsys/fcb/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+tests:
+  - test:
```
should not be a sequence, needs to be a map.

SebastianBoe commented on drivers/crc/crc8_sw.c:
```diff
@@ -0,0 +1,31 @@
+		val = (val << 4) ^ crc8_ccitt_small_table[val >> 4];
+	}
+	return val;
+}
```
Shouldn't this be lib/crc/crc8_sw.c?

SebastianBoe commented:
> Should I provide better way for providing user's custom flash map description instead of CONFIG_FLASH_MAP_CUSTOM option.

Perhaps allowing Kconfig'uration of a handful of predefined 'app' areas so the user can add his own areas (when he has a simple use-case with only a handful of areas) purely from Kconfig. This might
be more maintainable for some users since they won't have to copy-and-modify default_flash_map.

I don't really understand the use-case here though, so maybe I am way off :)

```
#ifdef CONFIG_FLASH_MAP_APP_AREA_0
	{
		.fa_id = 5,
		.fa_device_id = SOC_FLASH_0_ID,
		.fa_off  = CONFIG_FLASH_AREA_APP_0_OFFSET,
		.fa_size = CONFIG_FLASH_AREA_APP_0_SIZE,
	},
#endif
#ifdef CONFIG_FLASH_MAP_APP_AREA_1
	{
		.fa_id = 5,
		.fa_device_id = SOC_FLASH_0_ID,
		.fa_off  = CONFIG_FLASH_AREA_APP_1_OFFSET,
		.fa_size = CONFIG_FLASH_AREA_APP_1_SIZE,
	},
#endif

```

carlescufi commented:
> Should FCB module been moved to File System directory (so Kconfig as well)?
Yes, I think you should place it in `subsys/fs/fcb`

nashif commented on tests/subsys/storage/flash_mpa/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+tests:
+  - test:
```
still wrong

nvlsianpu committed:
tests: unit: lib: crc: Add test for crc8-ccitt implementation

CRC test suite was extended by crc8-ccitt test.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@updated

nashif commented on include/flash_map/flash_map.h:
```diff
@@ -0,0 +1,90 @@
+/*
```
can we put this in include/<subsystem>/flash_map.h
same for fcb.h

nvlsianpu commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -0,0 +1,270 @@
+	return true;
+}
+
+int flash_area_get_sectors(int idx, uint32_t *cnt, struct flash_sector *ret)
```
@mbolivar Should I add Linaro to copy right list? I mean that I take this form mcuboot, and it was your code.

mbolivar replied:
I think the main thing to do is preserve any copyright notices that were in the files which you're copying content from, so if there's Linaro headers in the original file, please make sure they're reproduced here, yes.

nvlsianpu replied:
So I will keep this as is - As mcuboot file from which i take this code doesn't have copyright list.

nvlsianpu commented:
@SebastianBoe Thanks from Idea for description customization. As so far I didn't see real user case too I think I will be better to improve this when something will appear (so not in this patch).

SebastianBoe commented on subsys/CMakeLists.txt:
```diff
@@ -10,3 +10,5 @@ add_subdirectory_ifdef(CONFIG_MCUBOOT_IMG_MANAGER  dfu)
 add_subdirectory_ifdef(CONFIG_USB                  usb)
 add_subdirectory(random)
+add_subdirectory(storage)
+add_subdirectory_ifdef(CONFIG_FCB  fs/fcb)
```
Have you kept this out of the fs directory because FCB does not depend on CONFIG_FILE_SYSTEM?

We need to find some way of getting it into fs as users will expect this code to be in there. If FCB does not require CONFIG_FILE_SYSTEM you will need to refactor away the ifdef FILE_SYSTEM in this file. PM me if I am being unclear.

nvlsianpu replied:
>Have you kept this out of the fs directory because FCB does not depend on CONFIG_FILE_SYSTEM?

Yes, exactly.

nvlsianpu replied:
FCB code do not depend on CONFIG_FILE_SYSTEM already.

carlescufi commented on include/fcb.h:
```diff
@@ -0,0 +1,134 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __SYS_FCB_H_
```
Why `__SYS_FCB_H` and not `_FCB_H_`?

carlescufi commented on include/fcb.h:
```diff
@@ -0,0 +1,134 @@
+}
+#endif
+
+#endif /* __SYS_FLASHVAR_H_ */
```
wrong comment

carlescufi commented on subsys/fs/fcb/fcb_priv.h:
```diff
@@ -0,0 +1,53 @@
+}
+#endif
+
+#endif
```
missing comment

carlescufi reviewed (changes requested):
Really nice work, just a couple of nitpicks

nvlsianpu commented:
@carlescufi updated

carlescufi commented:
@mbolivar can you take another look at this when you have time?

mbolivar commented:
> @mbolivar can you take another look at this when you have time?

@carlescufi I would actually prefer to ask @mike-scott to take a look at this. He's the person in our team that is most impacted

grandcat commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -0,0 +1,270 @@
+
+struct driver_map_entry {
+	u8_t id;
+	const char * const name;
```
When implementing multiple flash support [MCUB-75](https://runtimeco.atlassian.net/projects/MCUB/issues/MCUB-75) for mcuboot,  I went a similar approach as the `driver_map_entry`, but included the a pointer to `struct device` itself.
It helps to avoid many calls to `device_get_binding(..)`. Instead, it is called once and cached.

Could this be an option?

grandcat commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -0,0 +1,270 @@
+	/* nothing to do for now */
+}
+
+static struct device *get_flah_dev_from_id(u8_t id)
```
Typo: `get_flah_dev_from_id` -> `get_flash_dev_from_id`

grandcat commented on tests/subsys/storage/flash_mpa/src/main.c:
```diff
@@ -0,0 +1,103 @@
+/*
```
Typo in directory:
` tests/subsys/storage/flash_mpa` -> ` tests/subsys/storage/flash_map`

nvlsianpu committed:
subsys: storage: flash_map: devices bindings fetch optimalization

Moved fetch of flash device bindings to early initialization of the
application.
Device bindings are constant while the application is running so
it is better to fetch it at startup, and not every time flash_map
procedures are called.


Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
